src/objective/rank_obj.cu

```diff
@@ -601,26 +651,127 @@ struct MAPLambdaWeightComputer {
+    // label from the unsorted labels (from the original label list)
+
+    // For each item in the group, compute its MAP stats.
+    // Interleave the computation of map stats amongst different groups.
+
+    // First, determine postive labels in the dataset individually
+    auto nitems = segment_label_sorter.GetNumItems();
+    dh::caching_device_vector<uint32_t> dhits(nitems, 0);
+    // Original positions of the predictions after they have been sorted
+    const uint32_t *pred_original_pos = this->GetPredictionSorter().GetOriginalPositionsPtr();
+    // Unsorted labels
+    const float *unsorted_labels = dlabels;
+    auto DeterminePositiveLabelLambda = [=] __device__(uint32_t idx) {
+      return (unsorted_labels[pred_original_pos[idx]] > 0.0f) ? 1 : 0;
+    };  // NOLINT
+
+    thrust::transform(thrust::make_counting_iterator(static_cast<uint32_t>(0)),
+                      thrust::make_counting_iterator(nitems),
+                      dhits.begin(),
+                      DeterminePositiveLabelLambda);
+
+    // Next, prefix scan the positive labels that are segmented to accumulate them.
+    // This is required for computing the accumulated precisions
+    const auto &group_segments = segment_label_sorter.GetGroupSegments();
+    // Data segmented into different groups...
+    thrust::inclusive_scan_by_key(group_segments.begin(), group_segments.end(),
+                                  dhits.begin(),  // Input value
+                                  dhits.begin());  // In-place scan
+
+    // Compute accumulated precisions for each item, assuming positive and
+    // negative instances are missing.
+    // But first, compute individual item precisions
+    const auto *dgidx_arr = group_segments.data().get();
+    const auto *dhits_arr = dhits.data().get();
+    // Group info on device
+    const uint32_t *dgroups = segment_label_sorter.GetGroupsPtr();
+    uint32_t ngroups = segment_label_sorter.GetNumGroups();
+    auto ComputeItemPrecisionLambda = [=] __device__(uint32_t idx) {
+      if (unsorted_labels[pred_original_pos[idx]] > 0.0f) {
+        auto idx_within_group = (idx - dgroups[dgidx_arr[idx]]) + 1;
+        return MAPStats(static_cast<float>(dhits_arr[idx]) / idx_within_group,
+                        static_cast<float>(dhits_arr[idx] - 1) / idx_within_group,
+                        static_cast<float>(dhits_arr[idx] + 1) / idx_within_group,
+                        1.0f);
+      }
+      return MAPStats();
+    };  // NOLINT
+
+    thrust::transform(thrust::make_counting_iterator(static_cast<uint32_t>(0)),
+                      thrust::make_counting_iterator(nitems),
+                      this->dmap_stats_.begin(),
+                      ComputeItemPrecisionLambda);
+
+    // Lastly, compute the accumulated precisions for all the items segmented by groups.
+    // The precisions are accumulated within each group
+    thrust::inclusive_scan_by_key(group_segments.begin(), group_segments.end(),
```

Comment:
`thrust::inclusive_scan_by_key` allocates memory, so benefits from using a temporary allocator.

Reply:
done, although the performance barely budged. there were ~ 200 more allocation this entailed while performing this.

Reply:
Ok, I still want to make it a habit to use temporary allocators consistently with thrust.